### DISCUSSION TITLE
Outerjoin using nested JSON attribute not working

### DISCUSSION THREAD

**Discussion by creatixdevelopers** [sqlalchemy/sqlalchemy#11197] [Usage Questions] [answered]
I want to perform an outer join on  2 tables based on the value of a nested JSON attribute in first table, but the expected result is not what I expected. 
Here's some sample code to demonstrate the problem:
```
from sqlalchemy import create_engine, select, Integer, JSON, String
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Session
from sqlalchemy.orm import mapped_column, DeclarativeBase, Mapped

JSONVariant = JSON().with_variant(JSONB(), "postgresql")


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"
    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[dict] = mapped_column(JSONVariant, index=True, nullable=True)


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    color: Mapped[str] = mapped_column(String(16))
    value: Mapped[int] = mapped_column(Integer)


engine = create_engine("sqlite:///test.db")
Base.metadata.create_all(engine)

with Session(engine) as session:
    red = A(data={"color": "red"})
    blue = A(data={"color": "blue"})
    green = A(data={"color": "green"})

    a = B(color="red", value=10)
    b = B(color="blue", value=20)

    session.add_all([red, blue, green, a, b])
    session.commit()

    query = select(A.id, B.value).outerjoin(B, A.data["color"] == B.color)
    result = session.execute(query)
    print([r for r in result])

    # result = [(1, None), (2, None), (3, None)]
    # expected result = [(1, 10), (2, 20), (3, None)]
```
How can I modify the code to get the expected result?
Also, I use SQLite for development and PostgreSQL for production. Is the `JSONVariant` column type sufficient for such a use case or are there any issues I might run into?

**Comment by CaselIT** [comment]
Hi,

What's the sql that's being generated? My may need to cast the json data to string in postgresql

**Comment by creatixdevelopers** [comment, marked as answer]
Hi @CaselIT,

With sqlite I get:
`SELECT a.id, b.value FROM a LEFT OUTER JOIN b ON JSON_QUOTE(JSON_EXTRACT(a.data, ?)) = b.color`

And with postgresql I get:
`SELECT a.id, b.value FROM a LEFT OUTER JOIN b ON (a.data -> %(data_1)s) = b.color`

**Reply by CaselIT** [reply]
Seems like you need a cast at least in postgresql

**Reply by CaselIT** [reply]
Or the use of as_text() that IIRC should use the pg `->>` operator

**Reply by creatixdevelopers** [reply]
I got an error with `as_text()` but that prompted me to look for alternatives and I found `as_string()` which seems to work (with SQLite at least). So the query that ended up working was: 
`select(A.id, B.value).outerjoin(B, A.data["color"].as_string() == B.color`

**Reply by CaselIT** [reply]
I meant https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.JSON.Comparator.astext but I think that it may be only for the postgresql type.

As string should work through
